### PR TITLE
DB-9001 Conditionally purge tombstone during flush and minor compaction

### PR DESCRIPTION
**Pull request by arnaud-splice**

### CODE REVIEW

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST FAILED

**Comment by arnaud-splice** [comment]
jenkins please test branch @cdh5.12.2

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

File: hbase_storage/src/main/java/com/splicemachine/storage/HCell.java [hunk 1]
```diff
@@ -234,17 +240,23 @@ public long familyLength(){
     /* ****************************************************************************************************************/
     /*private helper methods*/
     private CellType parseType(Cell cell){
```
[the remark this answers is not in the document]
Comment (arnaud-splice) [reply hbase_storage/src/main/java/com/splicemachine/storage/HCell.java]: Yes, removing duplicate

File: hbase_storage/src/main/java/com/splicemachine/si/impl/server/SICompactionStateMutate.java [hunk 2]
```diff
@@ -88,19 +100,48 @@ private void mutate(Cell element, TxnView txn) throws IOException {
...
+                return firstWriteToken && deleteRightAfterFirstWriteTimestamp == maxTombstoneTimestamp;
+            case IF_FIRST_WRITE_PRESENT:
+                return firstWriteToken;
+        }
+        assert false: "Developer error";
+        return false;
+    }
+
+    private void removeDeletedRows() {
         SortedSet<Cell> cp = (SortedSet<Cell>)((TreeSet<Cell>)dataToReturn).clone();
```
[the remark this answers is not in the document]
Comment (arnaud-splice) [reply hbase_storage/src/main/java/com/splicemachine/si/impl/server/SICompactionStateMutate.java]: Done

File: hbase_storage/src/main/java/com/splicemachine/si/impl/server/SICompactionStateMutate.java [hunk 3]
```diff
@@ -88,19 +100,48 @@ private void mutate(Cell element, TxnView txn) throws IOException {
...
+    private boolean shouldRemoveMostRecentTombstone() {
+        switch (purgeConfig.getPurgeLatestTombstone()) {
+            case ALWAYS:
+                return true;
+            case IF_DELETE_FOLLOWS_FIRST_WRITE:
+                return firstWriteToken && deleteRightAfterFirstWriteTimestamp == maxTombstoneTimestamp;
+            case IF_FIRST_WRITE_PRESENT:
+                return firstWriteToken;
+        }
+        assert false: "Developer error";
```
[the remark this answers is not in the document]
Comment (arnaud-splice) [reply hbase_storage/src/main/java/com/splicemachine/si/impl/server/SICompactionStateMutate.java]: I'd rather leave this assert false here. This should not be reached. I can remove the message if you want though.

[the remark this answers is not in the document]
Comment (dgomezferro) [reply hbase_storage/src/main/java/com/splicemachine/si/impl/server/SICompactionStateMutate.java]: That makes sense, I misunderstood the direction of the assert.

File: splice_si_api/src/main/java/com/splicemachine/si/constants/SIConstants.java [hunk 4]
```diff
@@ -56,13 +56,17 @@
...
      * 7 = encoded user data
+     * 8 = token placed the first time we write to this row
      * 9 = column for causing write conflicts between concurrent transactions writing to parent and child FK tables
      */
-    public static final byte[] SNAPSHOT_ISOLATION_COMMIT_TIMESTAMP_COLUMN_BYTES = Bytes.toBytes("0");
-    public static final byte[] SNAPSHOT_ISOLATION_TOMBSTONE_COLUMN_BYTES = Bytes.toBytes("1");
-    public static final byte[] SNAPSHOT_ISOLATION_FK_COUNTER_COLUMN_BYTES = Bytes.toBytes("9");
-    public static final byte[] SNAPSHOT_ISOLATION_ANTI_TOMBSTONE_VALUE_BYTES = Bytes.toBytes("0");
-
+    public static final byte[] TIMESTAMP_COLUMN_BYTES = Bytes.toBytes("0");
```
Comment (msirek) [inline splice_si_api/src/main/java/com/splicemachine/si/constants/SIConstants.java]: This name shortening loses some of the description of what type of timestamp this is.  Maybe COMMIT_TIMESTAMP_COLUMN_BYTES would be better.

**Comment by changli6** [comment]
@arnaud-splice There are some conflicts, could you resolve it?

File: splice_si_api/src/main/java/com/splicemachine/si/constants/SIConstants.java (hunk 4 above)
Comment (arnaud-splice) [reply splice_si_api/src/main/java/com/splicemachine/si/constants/SIConstants.java]: Done.

**Pull request merged**